Scripts/Runtime/Core/Util/MathExtension.cs

```diff
@@ -0,0 +1,406 @@
+        /// <see cref="MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE"/>.
+        /// </returns>
+        /// <remarks>
+        /// Resolves infinite and NaN comparisons as false with improved performance vs <see cref="IsApproximately"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool4x4 IsApproximately(this float4x4 a, float4x4 b)
+        {
+            return new bool4x4(
+                a.c0.IsApproximately(b.c0),
+                a.c1.IsApproximately(b.c1),
+                a.c2.IsApproximately(b.c2),
+                a.c3.IsApproximately(b.c3)
+            );
+        }
+
+        /// <summary>
+        /// Checks if two <see cref="float"/> values are approximately equal according to the
+        /// <see cref="MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE"/>.
+        /// </summary>
+        /// <param name="a">The <see cref="float"/> to compare.</param>
+        /// <param name="b">The <see cref="float"/> to compare.</param>
+        /// <returns>
+        /// True if the delta between <see cref="a"/> and <see cref="b"/> is less than
+        /// <see cref="MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE"/>.
+        /// </returns>
+        /// <remarks>
+        /// Resolves infinite and NaN comparisons as true with the same logic as <see cref="IsEqualOrNaN"/> at
+        /// additional cost vs <see cref="IsApproximately"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsApproximatelySafe(this float a, float b)
+        {
+            return (math.abs(a - b) < MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE)
+                   | (math.isnan(a) & math.isnan(b))
+                   | (math.isinf(a) & math.isinf(b) & a == b);
+        }
+
+        /// <summary>
+        /// Checks if two <see cref="float2"/> values are approximately equal according to the
+        /// <see cref="MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE"/>.
+        /// </summary>
+        /// <param name="a">The <see cref="float2"/> to compare.</param>
+        /// <param name="b">The <see cref="float2"/> to compare.</param>
+        /// <returns>
+        /// True (component-wise) if the delta between <see cref="a"/> and <see cref="b"/> is less than
+        /// <see cref="MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE"/>.
+        /// </returns>
+        /// <remarks>
+        /// Resolves infinite and NaN comparisons as true with the same logic as <see cref="IsEqualOrNaN"/> at
+        /// additional cost vs <see cref="IsApproximately"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool2 IsApproximatelySafe(this float2 a, float2 b)
+        {
+            return (math.abs(a - b) < MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE)
+                   | (math.isnan(a) & math.isnan(b))
+                   | (math.isinf(a) & math.isinf(b) & a == b);
```

Comment:
Do we run into issues when only some of the elements are NaN or Infinity? 

`float2 a = new float2(5.0f, NaN)`
`float2 b = new float2(5.0f, NaN)`

We do the first check which should give... `(0.0f, NaN) < FLOATING_POINT_EQUALITY_TOLERANCE`? 
Which I think will return false because of the NaN? 

Then we do the `isnan` check but that checks both elements and 5.0f isn't NaN so they both fail. 

The above example should be true, but will return false? I wonder if instead you need to access the array indexer and run each component individually through the single `IsApproximatelySafe` call?

Reply:
Broke it down step by step for you below. Keep in mind that all of the operations are component-wise and we get a `bool2` out of this.

``` csharp
a-b   // 0, float.NaN
a-b < MathUtil.FLOATING_POINT_EQUALITY_TOLERANCE   // true, false
math.isnan(a) & math.isnan(b)   // false, true
(math.isinf(a) & math.isinf(b) & a == b)   // false, false

(true, false) | (false, true) | (false, false)   // true, true
```

I added a unit test for float2, float3, and tweaked float 4 to cover this. It was a good case!

Reply:
Ok got it, it makes sense now.